File: tests/libzstd/tests/tagged-port/safe/tests/layout/public_types.c
#include <stddef.h>
#include <stdio.h>

#include "zstd.h"

#define PRINT_LAYOUT(type_name) \
    printf("%s size=%zu align=%zu\n", #type_name, sizeof(type_name), _Alignof(type_name))

#define PRINT_POINTER_LAYOUT(label, pointee_type) \
    printf("%s* size=%zu align=%zu\n", label, sizeof(pointee_type*), _Alignof(pointee_type*))

int main(void) {
    PRINT_LAYOUT(ZSTD_inBuffer);
    PRINT_LAYOUT(ZSTD_outBuffer);
    PRINT_LAYOUT(ZSTD_customMem);
    PRINT_LAYOUT(ZSTD_frameHeader);
    PRINT_LAYOUT(ZSTD_Sequence);
    PRINT_LAYOUT(ZSTD_bounds);
    PRINT_LAYOUT(ZSTD_frameProgression);

    PRINT_POINTER_LAYOUT("ZSTD_CCtx", ZSTD_CCtx);
    PRINT_POINTER_LAYOUT("ZSTD_DCtx", ZSTD_DCtx);
    PRINT_POINTER_LAYOUT("ZSTD_CDict", ZSTD_CDict);
    PRINT_POINTER_LAYOUT("ZSTD_DDict", ZSTD_DDict);
    PRINT_POINTER_LAYOUT("ZSTD_CCtx_params", ZSTD_CCtx_params);
    PRINT_POINTER_LAYOUT("ZSTD_threadPool", ZSTD_threadPool);
    return 0;
}

File: tests/libzstd/tests/tagged-port/safe/tests/layout/public_types.rs
#![allow(non_camel_case_types)]

use std::mem::{align_of, size_of};

use zstd::ffi::types::{
    ZSTD_CCtx, ZSTD_CCtx_params, ZSTD_CDict, ZSTD_DCtx, ZSTD_DDict, ZSTD_Sequence,
    ZSTD_bounds, ZSTD_customMem, ZSTD_frameHeader, ZSTD_frameProgression, ZSTD_inBuffer,
    ZSTD_outBuffer, ZSTD_threadPool,
};

fn emit<T>(name: &str) {
    println!("{name} size={} align={}", size_of::<T>(), align_of::<T>());
}

fn main() {
    emit::<ZSTD_inBuffer>("ZSTD_inBuffer");
    emit::<ZSTD_outBuffer>("ZSTD_outBuffer");
    emit::<ZSTD_customMem>("ZSTD_customMem");
    emit::<ZSTD_frameHeader>("ZSTD_frameHeader");
    emit::<ZSTD_Sequence>("ZSTD_Sequence");
    emit::<ZSTD_bounds>("ZSTD_bounds");
    emit::<ZSTD_frameProgression>("ZSTD_frameProgression");

    emit::<*mut ZSTD_CCtx>("ZSTD_CCtx*");
    emit::<*mut ZSTD_DCtx>("ZSTD_DCtx*");
    emit::<*mut ZSTD_CDict>("ZSTD_CDict*");
    emit::<*mut ZSTD_DDict>("ZSTD_DDict*");
    emit::<*mut ZSTD_CCtx_params>("ZSTD_CCtx_params*");
    emit::<*mut ZSTD_threadPool>("ZSTD_threadPool*");
}
